pkg/clients/service.go

```diff
@@ -0,0 +1,196 @@
+	hClient "github.com/ory/hydra-client-go/v2"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type PaginationLinksResponse struct {
+	First string `json:"first,omitempty"`
+	Last  string `json:"last,omitempty"`
+	Prev  string `json:"prev,omitempty"`
+	Next  string `json:"next,omitempty"`
+}
+
+type PaginationMetadata struct {
+	Links      *PaginationLinksResponse
+	TotalCount int
+}
+
+type Service struct {
+	hydra HydraClientInterface
+
+	tracer  trace.Tracer
+	monitor monitoring.MonitorInterface
+	logger  logging.LoggerInterface
+}
+
+func (s *Service) parseServiceError(r *http.Response, e error) (*hClient.ErrorOAuth2, error) {
+	// The hydra client does not return any errors, we need to parse the response body and create our
+	// own objects.
+	// Should we use our objects instead of reusing the ones from the sdk?
+	se := hClient.NewErrorOAuth2()
+
+	if r == nil {
+		s.logger.Debugf("Got no response from hydra service")
+		se.SetError("internal_server_error")
+		se.SetErrorDescription("Failed to call hydra service")
+		se.SetStatusCode(http.StatusInternalServerError)
+		return se, nil
+	}
+
+	json_data, err := io.ReadAll(r.Body)
+	if err != nil {
+		s.logger.Debugf("Failed to read response body: %s", err)
+		return se, err
+	}
+	err = json.Unmarshal(json_data, se)
+	if err != nil {
+		s.logger.Debugf("Failed to unmarshal JSON: %s", err)
+		return se, err
+	}
+	se.SetStatusCode(int64(r.StatusCode))
+
+	return se, nil
+}
+
+func (s *Service) GetClient(ctx context.Context, clientID string) (*hClient.OAuth2Client, *hClient.ErrorOAuth2, error) {
+	_, span := s.tracer.Start(ctx, "hydra.OAuth2Api.GetOAuth2Client")
+	defer span.End()
+
+	c, resp, err := s.hydra.OAuth2Api().
+		GetOAuth2Client(ctx, clientID).
+		Execute()
+
+	if err != nil {
+		se, err := s.parseServiceError(resp, err)
+		return nil, se, err
+	}
+	return c, nil, nil
+}
+
+func (s *Service) CreateClient(ctx context.Context, client *hClient.OAuth2Client) (*hClient.OAuth2Client, *hClient.ErrorOAuth2, error) {
+	_, span := s.tracer.Start(ctx, "hydra.OAuth2Api.CreateOAuth2Client")
+	defer span.End()
+
+	c, resp, err := s.hydra.OAuth2Api().
+		CreateOAuth2Client(ctx).
+		OAuth2Client(*client).
+		Execute()
+
+	if err != nil {
+		se, err := s.parseServiceError(resp, err)
+		return nil, se, err
+	}
+	return c, nil, nil
+}
+
+func (s *Service) UpdateClient(ctx context.Context, client *hClient.OAuth2Client) (*hClient.OAuth2Client, *hClient.ErrorOAuth2, error) {
+	_, span := s.tracer.Start(ctx, "hydra.OAuth2Api.SetOAuth2Client")
+	defer span.End()
+
+	c, resp, err := s.hydra.OAuth2Api().
+		SetOAuth2Client(ctx, *client.ClientId).
+		OAuth2Client(*client).
+		Execute()
+
+	if err != nil {
+		se, err := s.parseServiceError(resp, err)
+		return nil, se, err
+	}
+	return c, nil, nil
+}
+
+func (s *Service) ListClients(ctx context.Context, cName string, owner string, ps int64, pt string, u *url.URL) ([]hClient.OAuth2Client, *PaginationMetadata, *hClient.ErrorOAuth2, error) {
```

Comment:
I think a pagination struct composed with a specific service pagination struct would work better here instead of all the arts (see golang composition)


```suggestion
func (s *Service) ListClients(ctx context.Context, cName, owner, pt string, ps int64, u *url.URL) ([]hClient.OAuth2Client, *PaginationMetadata, *hClient.ErrorOAuth2, error) {
```